src/gpu/intel/jit/ir/core.hpp

```diff
@@ -658,10 +595,72 @@ class object_t;
 class ir_mutator_t;
 class ir_visitor_t;
 
+enum class ir_type_id_t : uint8_t {
+    undef = 0,
+
+#define HANDLE_IR_OBJECT(type) type,
+
+    // Create typeid for objects which can be visited/mutated. These need to be
+    // first as the typeid is used as an index into an array to dispatch to the
+    // correct mutate function.
+    HANDLE_CORE_IR_OBJECTS()
+
+#undef HANDLE_IR_OBJECT
+
+};
+
+constexpr bool is_expr(ir_type_id_t id) {
+    switch (id) {
+#define HANDLE_IR_OBJECT(type) \
+    case ir_type_id_t::type: return true;
+
+        HANDLE_EXPR_IR_OBJECTS()
+
+#undef HANDLE_IR_OBJECT
+
+        default: return false;
+    }
+}
+
+constexpr bool is_stmt(ir_type_id_t id) {
+    switch (id) {
+#define HANDLE_IR_OBJECT(type) \
+    case ir_type_id_t::type: return true;
+
+        HANDLE_STMT_IR_OBJECTS()
+
+#undef HANDLE_IR_OBJECT
+
+        default: return false;
+    }
+}
+
+struct type_info_t {
+    constexpr type_info_t(ir_type_id_t type_id, const char *name)
+        : type_id(type_id)
+        , name(name)
+        , is_expr(jit::is_expr(type_id))
+        , is_stmt(jit::is_stmt(type_id)) {}
+
+    ir_type_id_t type_id = ir_type_id_t::undef;
+    const char *name = nullptr;
+    bool is_expr = false;
+    bool is_stmt = false;
+
+    bool operator==(const type_info_t &other) const {
+        if (type_id == ir_type_id_t::undef
+                && other.type_id == ir_type_id_t::undef)
+            return (name == other.name) || std::strcmp(name, other.name) == 0;
```

Comment:
I don't think using a string comparison is a good idea here. Is there anything stopping us from just using [std::type_info](https://en.cppreference.com/w/cpp/types/type_info.html)?

Reply:
Thanks, changed. I don't see overhead from `std::type_info`, and if no issues with compiler/OS - then we should be good to use it.